src/coord/src/coord.rs

```diff
@@ -1437,15 +1443,32 @@ impl<S: Append + 'static> Coordinator<S> {
         }
     }
 
-    /// Instruct the dataflow layer to cancel any ongoing, interactive work for
-    /// the named `conn_id`.
+    /// Handles a cancel request from a client (e.g. ctrl+C).
     async fn handle_cancel(&mut self, conn_id: u32, secret_key: u32) {
+        self.cancel_inner(conn_id, Some(secret_key), true).await
+    }
+
+    /// Handles cancelling on going work in the face of an error, such as a timeout.
+    async fn force_cancel_cleanup(&mut self, conn_id: u32) {
+        self.cancel_inner(conn_id, None, false).await
+    }
+
+    /// Cancel any ongoing, interactive work for the named `conn_id`.
+    ///
+    /// # Parameters
+    /// - `secret_key` must be `Some` when handling client-initiated cancels,
+    /// but can be `None` to permit the system to force a cancellation.
+    /// - `send_response` must be false when dealing with a forced cancellation
+    ///   because the client is already set to receive another response.
+    async fn cancel_inner(&mut self, conn_id: u32, secret_key: Option<u32>, send_response: bool) {
```

Comment:
Can the secret key checking be handled by `handle_cancel` so that this doesn't have to know about it?